In [1]:
import pandas as pd

# General Description

In our study, you will be asked to finish three exploratory data analysis tasks using GitHub Copilot. In each task, the number of quries/comments you can use is different. 

### Before you begin, attention!

- You can only use comment and auto-fill functions of Github Copilot to solve the task. 
- You can also fix small bugs like wrong variable name or wrong list index. 
- You can change the comment to generate new code as well.
- You cannot, however, write codes by yourself instead of using comment to generate. If you are not sure whether your action is permitted, please just ask us.
- Due to study requirement, we can only answer questions about the problem setting. We cannot answer questions about how to actually solve the problem. We may show you our solutions at the end of the task.
- The time limit for each part is **15 minutes**. If you believe that you have already finished the task before the time limit, or you feel that spending more time will not help you solve the problem, you can just tell us and move on. There is no need to worry about that.
- We encourage you to **think aloud** during the programming process, which will help us to know what users are thinking when using GitHub Copilot.

# Task 1

### Load data

In [2]:
df1 = pd.read_csv('equity_value_data.csv')
df1.head()

timestamp  close_equity                           user_id
0  2016-11-16T00:00:00Z         48.16  bcef4fa9b0bdf22bcf7deae708decf03
1  2016-11-17T00:00:00Z         48.16  bcef4fa9b0bdf22bcf7deae708decf03
2  2016-11-18T00:00:00Z         48.16  bcef4fa9b0bdf22bcf7deae708decf03
3  2016-11-21T00:00:00Z         48.16  bcef4fa9b0bdf22bcf7deae708decf03
4  2016-11-22T00:00:00Z         48.16  bcef4fa9b0bdf22bcf7deae708decf03

### Task Description

This data set contains information for investors with their equity in an online trading platform A. You can see that there are three columns in the data set. The `timestamp` indicates the when the record is written. The `close_equity` means the close equity of that user in a particular day. The `user_id` is the unique identifier of the user. Note that this data set **only records timestamps with `close_equity > 10`**, and thus you can find that the timestamps are not consecutive for some users. We call the missing timestamp period **`inactive period`**. In this task, you need to find out the percentage of users who have a consecutive `inactive period` longer than 28 days.

### Example
| timestamp            | close_equity | user_id                          |
|----------------------|--------------|----------------------------------|
| 2016-10-16T00:00:00Z | 48.16        | bcef4fa9b0bdf22bcf7deae708decf03 |
| 2016-10-17T00:00:00Z | 48.16        | bcef4fa9b0bdf22bcf7deae708decf03 |
| 2016-12-29T00:00:00Z | 48.16        | bcef4fa9b0bdf22bcf7deae708decf03 |
| 2016-10-16T00:00:00Z | 22.34        | 3276897f94d72719b16f1648f7166384 |
| 2016-10-17T00:00:00Z | 25.68        | 3276897f94d72719b16f1648f7166384 |
| 2016-11-01T00:00:00Z | 34.22        | 3276897f94d72719b16f1648f7166384 |

The answer for this example is 50%.

In this table, the user with id `bcef4fa9b0bdf22bcf7deae708decf03` has a consecutive inactive period longer than 28 days between `2016-10-17` to `2016-12-29`. While the user with id `3276897f94d72719b16f1648f7166384` does not have a consecutive inactive period longer than 28 days. 

### Assume that you are not familiar with data analytics at all, and you can only describe the question to Copilot and hope it can help you out. Now, please try to solve the problem by just using **one** prompt/comment (can be multiple sentences) and let the Github Copilot finish the rest of work. Do not be afraid if you fail.

# Task 2

### Load Data

In [4]:
df2 = pd.read_csv('International_Report_Departures.csv')
df2.head()

data_dte  Year  Month  usg_apt_id usg_apt  usg_wac  fg_apt_id fg_apt  \
0  05/01/2006  2006      5       12016     GUM        5      13162    MAJ   
1  04/01/2009  2009      4       10397     ATL       34      16321    ZRH   
2  05/01/2003  2003      5       10299     ANC        1      13856    OKO   
3  02/01/2005  2005      2       11057     CLT       36      14849    SKB   
4  01/01/2004  2004      1       12478     JFK       22      11625    EZE   

   fg_wac  airlineid carrier  carriergroup        type  Scheduled  Charter  \
0     844      20177     PFQ             1  Departures          0       10   
1     486      19790      DL             1  Departures         54        0   
2     736      20007      5Y             1  Departures          0       15   
3     275      20355      US             1  Departures          8        0   
4     303      19535      AR             0  Departures         30        0   

   Total  
0     10  
1     54  
2     15  
3      8  
4     30

### Task Description
You can see that there are lots of departures records with a variety of fields. We are particularly interested in the `usg_apt` field. This field represents the departure airport in the US. We want to find out the top 10 airport each year from 2019 to 2022 based on number of departures records. And then plot a bar plot with x-axis to be the airport, y-axis to be the number of times it enters top 10.

### Example

![Desired Output](./bar_plot.jpg)

### Assume that you are familiar with other data analysis tools like Excel, R or other programming languages. So you know some general steps to solve such problems. But you are not familiar with Python, which prevents you from further breakdown to specific codes. Now, try to break down the problem a little bit, using 3-5 different prompts/comments and let the Copilot finish the rest of work. Don't worry if you fail.

# Task 3

### Load Data

In [8]:
df3 = pd.read_csv('vgsales.csv')
df3.head()

Rank                      Name Platform    Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37

### Task Description

In this data set, there are records showing the sales of popular video games. We are interested in the `Platform` column, which indicates the platform of the game. We want to check whether multi-platform games have a higher `Global_Sales` than those single-platform games.

### Example

| Name                 | Platform     | Global_Sales                     |
|----------------------|--------------|----------------------------------|
| A                    | PC           | 80                               |
| B                    | PS4          | 65                               |
| B                    | PC           | 25                               |
| C                    | PS4          | 70                               |
| C                    | PC           | 10                               |

In this example, overall sales of `B` is 90, overall sales of `C` is 80. And the average multi-platform sales is indeed better on average than single-platform sales.


 ### Assume that you are quite familiar with data analysis using Python, and you know exactly what each step should do (which is likely for our participants). Now, try to break down the problem into as many sub-steps as you want. Writing corresponding prompts and letting the Copilot generate code for you. Don't worry if you fail.